scripts/pylib/twister/twisterlib/testplan.py

```diff
@@ -542,6 +542,24 @@ def get_platform(self, name):
                 break
         return selected_platform
 
+    def handle_quarantined_tests(self, instance: TestInstance, plat: Platform):
+        matched_quarantine = False
+        if self.quarantine:
+            matched_quarantine = self.quarantine.get_matched_quarantine(
+                instance.testsuite.id, plat.name, plat.arch, plat.simulation
+            )
+            if matched_quarantine:
+                if self.options.quarantine_verify and instance.reason:
+                    # resume case to be able to run
+                    instance.status = None
+                    instance.reason = None
```

Comment:
I don't understand why do you complicate this method and add this?
When you first run twister with a quarantine, then you would not find any artefacts for quarantined configurations, so you can not use `--quarantine-verify` with second run with `--test-only`.
The only scenario is when you first run twister without quarantine, then you have all artefacts in `twister-out` and there are no configurations marked in `twister.json` to be skipped by quarantine - so when you run twister second time with `--test-only` you can use quarantine file to skip problematic tests, or run with `--quarantine-verify` to run only those tests.

Reply:
I add this is because this scenario:
build with quarantine, and run with auarantine, and we want to retry the case them we need re-enable to run it. this is because in load_from_file, there is a processing code as below
```
               status = ts.get('status', None)
                reason = ts.get("reason", "Unknown")
                if status in ["error", "failed"]:
                    instance.status = None
                    instance.reason = None
                    instance.retries += 1
                # test marked as passed (built only) but can run when
                # --test-only is used. Reset status to capture new results.
                elif status == 'passed' and instance.run and self.options.test_only:
                    instance.status = None
                    instance.reason = None
                else:
                    instance.status = status
                    instance.reason = reason
```
so instead set the status and reason as `None`, they will still be filter out. so for this scenario, we need clear the status and reason when the quarantine is matched and we are in verify mode.

Reply:
Once again:
> When you first run twister with a quarantine, then you would not find any artefacts for quarantined configurations, so you can not use --quarantine-verify with second run with --test-only.

So, when you build and run with the quarantine enabled:
`scripts/twister -vv -p native_sim -T samples/hello_world --quarantine-list quarantine.yaml`
where quarantine.yaml is:
```
- scenarios:
  - sample.basic.helloworld
```
then test configuration is filtered, not built, so you wouldn't find any artefacts in `twister-out`,
and when you re-run it with quarantine-verify, you'd receive an error from twister:
```
scripts/twister -vv -p native_sim -T samples/hello_world --test-only --quarantine-list quarantine.yaml --quarantine-verify
FileNotFoundError: [Errno 2] No such file or directory: '<...>/zephyr/twister-out/native_sim/samples/hello_world/sample.basic.helloworld/zephyr/zephyr.exe'
```

Without modifying outputs between twister re-runs or copying artefacts from another builds, this scenario does not have any sense.
With that change you only provided a regression:
- quarantine matched + quarantine verified + another filter matched

Reply:
@gchwier, it is a valid scenario, let's see, in the beginning of testing, we do not know what to quarantine, so we build without the `--quarantine-list`, or build with at list that is not include the question caes.  but during the testing, we find there is a blocker case, and we need skip it, so we can modify the quarantine list and skip those cases, but they are build. and gradually enable those case to try out which one is the real blocker.

Reply:
> Without modifying outputs between twister re-runs or copying artefacts from another builds, this scenario does not have any sense. With that change you only provided a regression:
> 
> * quarantine matched + quarantine verified + another filter matched

and as to this regression, I am not clear. the test_quarantine.py test the quarantine list with --verify, and support to list all 7 quarantine cases, but it check 2? I do not know why.

Reply:
Quarantine feature should be simple. I agree for scenario, where you build all without the quarantine, and then you enable it (with `--test-only`), and it works perfectly now with your changes (without part `# resume case to be able to run`).
But if you want to change quarantine many times between re-running twister, than you should think about re-building, or to update twister.json manually - or do not use quarantine for such a scenario (filter tests in different way). 

And as I wrote in previous comment, you've added a regression with code:
```
    if self.options.quarantine_verify and instance.reason:
        # resume case to be able to run
        instance.status = None
        instance.reason = None
```
> I am not clear. the test_quarantine.py test the quarantine list with --verify, and suppose to list all 7 quarantined cases, but it checks 2? I do not know why.

If you use `--quarantine-verify` you want to run only tests, that are matched by quarantine rules, all others are filtered with a reason: "Not under quarantine".
So in that test, only 2 cases should be filtered! @LukaszMrugala maybe you explain it better.

Reply:
> But if you want to change quarantine many times between re-running twister, than you should think about re-building, or to update twister.json manually - or do not use quarantine for such a scenario (filter tests in different way).

But this will be additional efforts for this.

Reply:
Check documentation:
https://docs.zephyrproject.org/latest/develop/test/twister.html#quarantine
```
The current status of tests on the quarantine list can also be verified by adding --quarantine-verify to the above argument. This will make twister skip all tests which are not on the given list.
```
Blackbox tests are testing that feature. If you want to change it, you should also update this documentation (and tests).

Reply:
> If you use `--quarantine-verify` you want to run only tests, that are matched by quarantine rules, all others are filtered with a reason: "Not under quarantine".
> So in that test, only 2 cases should be filtered! @LukaszMrugala maybe you explain it better.

I think this is a bug. see below list, all of them are quarantined by list, but when using --verify should those case be able to run?

```
[
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert2'),
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert3'),
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert1'), 
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group1/subgroup2/dummy.agnostic.group1.subgroup2', 'dummy.agnostic.group1.subgroup2.assert'), 
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group1/subgroup1/dummy.agnostic.group1.subgroup1', 'dummy.agnostic.group1.subgroup1.assert'), 
('qemu_x86', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group1/subgroup1/dummy.agnostic.group1.subgroup1', 'dummy.agnostic.group1.subgroup1.assert'), 
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/device/group/dummy.device.group', 'dummy.device.group.assert')]
```

Reply:
> > But if you want to change quarantine many times between re-running twister, than you should think about re-building, or to update twister.json manually - or do not use quarantine for such a scenario (filter tests in different way).
> 
> But this will be additional efforts for this.

@hakehuang I see your point and I've checked your scenario once again, but IMO this is a narrow case, and can provide much more problems / regression issues.
Look:
* `apply_filters` does filtering of platforms, toolchains, architectures etc. And this method is not used in path with `load_from_file` (--test-only, --only-failed, --load-tests)
* if you clean status when the quarantine is matched, then you can miss other filters added with original run (regression)
* probable you can solve it with additional if-statements (check if the quarantine filter is the only one filter, if it is re-running of twister, to not break path with `apply_filters` ... etc.), but then that code will be much more complicated than it is now, and it will be really difficult to understand.

and once again, I agree to add quarantine filter to be checked in re-run path (e.g. with --test-only), I am only against to clean status and reason (that additional logic), because it provides regression

Reply:
> Blackbox tests are testing that feature. If you want to change it, you should also update this documentation (and tests).

@gchwier , the document is right. but the black testing seems not. As I mentioned above, 7 cases are in the quarantine list, and by adding --quarantine-verify only two of them can be run, why?

Reply:
why 2, not 7? Please test on the original code (or at least comment those lines with cleaning status and reason), and debug / print reasons, why some tests are filtered:
```
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group1/subgroup1/dummy.agnostic.group1.subgroup1', 'dummy.agnostic.group1.subgroup1.assert',
'Not in testsuite platform allow list')
('qemu_x86', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group1/subgroup1/dummy.agnostic.group1.subgroup1', 'dummy.agnostic.group1.subgroup1.assert',
None)
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group1/subgroup2/dummy.agnostic.group1.subgroup2', 'dummy.agnostic.group1.subgroup2.assert',
'Not in testsuite platform allow list')
('qemu_x86', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group1/subgroup2/dummy.agnostic.group1.subgroup2', 'dummy.agnostic.group1.subgroup2.assert',
'Not under quarantine')
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert3',
'Not in testsuite platform allow list')
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert2',
'Not in testsuite platform allow list')
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert1',
'Not in testsuite platform allow list')
('qemu_x86', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert3',
'Not under quarantine')
('qemu_x86', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert2',
'Not under quarantine')
('qemu_x86', 'scripts/tests/twister_blackbox/test_data/tests/dummy/agnostic/group2/dummy.agnostic.group2', 'dummy.agnostic.group2.assert1',
'Not under quarantine')
('frdm_k64f', 'scripts/tests/twister_blackbox/test_data/tests/dummy/device/group/dummy.device.group', 'dummy.device.group.assert',
None)
('qemu_x86', 'scripts/tests/twister_blackbox/test_data/tests/dummy/device/group/dummy.device.group', 'dummy.device.group.assert',
'Not under quarantine')
```

Reply:
oh, I miss check the testcase yaml file, so they are filter by `platform_allow`, Let me revert my stupid addons. for clarify. @gchwier